tls/s2n_x509_validator.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#pragma once
+
+#include "api/s2n.h"
+
+#include <openssl/x509v3.h>
+
+/** Return TRUE for trusted, FALSE for untrusted **/
+typedef uint8_t (*verify_host) (const char *host_name, size_t host_name_len, void *data);
+struct s2n_connection;
+
+/**
+ * Trust store simply contains the trust store each connection should validate certs against.
+ * For most use cases, you only need one of these per application.
+ */
+struct s2n_x509_trust_store {
+    X509_STORE *trust_store;
+};
+
+/**
+ * You should have one instance of this per connection.
+ */
+struct s2n_x509_validator {
+    struct s2n_x509_trust_store *trust_store;
+    STACK_OF(X509) *cert_chain;
+
+    uint8_t validate_certificates;
```

Comment:
More of a question to others, do we want to expose a public API to manually set `validate_certificates`? It would be useful for testing. Are we okay with letting people "shoot themselves in the foot" if they ask to do it?

Reply:
I think we have to for testing purposes. If I'm working on a client for example, I don't want to be fighting PKI when I'm just trying to get my http implementation working. So I normally turn it off, then come back and fix that when I'm done with my prototype.

Also, coorporate infrastructures often have a "special" way of deploying the trusted certs that might not be available at development time.

Reply:
The mechanism now is to call s2n_config_disable_x509_validation(). That's the only way to turn it off.